File: apps/android/app/src/main/java/com/example/creditcard/utils/CardMetrics.kt
package com.example.creditcard.utils

import com.example.creditcard.data.SharedCard
import java.math.BigDecimal
import java.math.RoundingMode
import java.util.Locale

/** Rules v1: keep independent regions/currencies apart and never invent an exchange rate. */
object CardMetrics {
    data class PoolKey(val country: String, val bank: String, val currency: String)
    private val parentheses = Regex("\\s*[（(][^（）()]*[）)]\\s*")
    private val whitespace = Regex("\\s+")
    fun bankKey(value: String): String = value.trim().replace(parentheses, "")
        .replace(whitespace, "").lowercase(Locale.ROOT)
    fun currency(card: SharedCard): String = card.type.trim().uppercase(Locale.ROOT)
    fun poolKey(card: SharedCard) = PoolKey(card.country.trim(), bankKey(card.bank), currency(card))
    private fun money(value: Double): BigDecimal = if (value.isFinite() && value >= 0)
        BigDecimal.valueOf(value).setScale(2, RoundingMode.HALF_UP) else BigDecimal.ZERO

    fun creditLimits(cards: List<SharedCard>): Map<String, Double> {
        val sums = linkedMapOf<String, BigDecimal>()
        val pools = linkedMapOf<PoolKey, BigDecimal>()
        for (card in cards) {
            if (card.cardCategory == "debit") continue
            val amount = money(card.limit)
            if (card.isSharedLimit) {
                val key = poolKey(card)
                pools[key] = (pools[key] ?: BigDecimal.ZERO).max(amount)
            } else sums[currency(card)] = (sums[currency(card)] ?: BigDecimal.ZERO) + amount
        }
        for ((key, amount) in pools) sums[key.currency] = (sums[key.currency] ?: BigDecimal.ZERO) + amount
        return sums.mapValues { it.value.toDouble() }.toSortedMap()
    }
}

File: apps/ios/Domain/CardMetrics.swift
import Foundation

/// Business rules v1; fixture contract lives in contracts/card-wallet. No exchange-rate conversion.
public enum CardMetrics {
    public struct PoolKey: Hashable, Sendable {
        public let country: String
        public let bank: String
        public let currency: String
    }

    public static func currency(_ card: SharedCard) -> String {
        (card.type ?? "").trimmingCharacters(in: .whitespacesAndNewlines).uppercased()
    }

    public static func poolKey(_ card: SharedCard) -> PoolKey {
        PoolKey(country: card.country.trimmingCharacters(in: .whitespacesAndNewlines),
                bank: BankNameNormalizer.normalizedKey(card.bank), currency: currency(card))
    }

    private static func money(_ value: Double?) -> Decimal {
        guard let value, value.isFinite, value >= 0,
              var decimal = Decimal(string: String(value), locale: Locale(identifier: "en_US_POSIX")) else { return 0 }
        var rounded = Decimal()
        NSDecimalRound(&rounded, &decimal, 2, .plain)
        return rounded
    }

    public static func creditLimits(cards: [SharedCard]) -> [String: Double] {
        var sums: [String: Decimal] = [:]
        var pools: [PoolKey: Decimal] = [:]
        for card in cards where card.cardCategory != "debit" {
            let amount = money(card.limit)
            if card.isSharedLimit {
                let key = poolKey(card)
                pools[key] = max(pools[key] ?? 0, amount)
            } else {
                sums[currency(card), default: 0] += amount
            }
        }
        for (key, amount) in pools { sums[key.currency, default: 0] += amount }
        return sums.mapValues { NSDecimalNumber(decimal: $0).doubleValue }
    }
}
